server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -116,4 +205,158 @@ public void collectSearchShards(String routing, IntConsumer consumer) {
+
+        ExtractFromSource(int routingNumShards, int routingFactor, String indexName, List<String> routingPaths) {
+            super(routingNumShards, routingFactor);
+            this.indexName = indexName;
+            this.include = FilterPath.compile(Set.copyOf(routingPaths));
+        }
+
+        @Override
+        public int indexShard(
+            boolean idIsAutoGenerated,
+            String id,
+            @Nullable String routing,
+            XContentType sourceType,
+            BytesReference source
+        ) {
+            if (idIsAutoGenerated == false) {
+                throw new IllegalArgumentException(error("indexing with a specified id"));
+            }
+            if (routing != null) {
+                throw new IllegalArgumentException(error("indexing with a specified routing"));
+            }
+            try {
+                try (
+                    XContentParser parser = sourceType.xContent()
+                        .createParser(
+                            NamedXContentRegistry.EMPTY,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                            source.streamInput(),
+                            include,
+                            null
+                        )
+                ) {
+                    parser.nextToken(); // Move to first token
+                    if (parser.currentToken() == null) {
+                        throw new IllegalArgumentException("Error extracting routing: source didn't contain any routing fields");
+                    }
+                    int hash = extractObject(parser);
+                    ensureExpectedToken(null, parser.nextToken(), parser);
+                    return hashToShardId(hash);
+                }
+            } catch (IOException | ParsingException e) {
+                throw new IllegalArgumentException("Error extracting routing: " + e.getMessage(), e);
+            }
+        }
+
+        private static int extractObject(XContentParser source) throws IOException {
+            ensureExpectedToken(Token.FIELD_NAME, source.nextToken(), source);
+            String firstFieldName = source.currentName();
+            source.nextToken();
+            int firstHash = extractItem(source);
+            if (source.currentToken() == Token.END_OBJECT) {
+                // Just one routing key in this object
+                // Use ^ like Map.Entery's hashcode
+                return Murmur3HashFunction.hash(firstFieldName) ^ firstHash;
+            }
+            String[] fields = new String[] { firstFieldName, null };
+            int[] hashes = new int[] { firstHash, 0 };
+            int length = 1;
+            do {
+                ensureExpectedToken(Token.FIELD_NAME, source.currentToken(), source);
+                String fieldName = source.currentName();
+                source.nextToken();
+                if (length >= fields.length) {
+                    fields = ArrayUtil.grow(fields, length + 1);
```

Comment:
Would we not want to grow by more than 1? This seems expensive for the cases where we have 5-10 fields that match.

Reply:
`length+1` is a minimum here. `grow` will oversize the array based on some heuristics about the jvm.